### DISCUSSION TITLE
Ending up with non-semver semver in 6.0-beta2

### DISCUSSION THREAD

**Discussion by NetTecture** [GitTools/GitVersion#3609] [Q&A] [answered]
**Describe the bug**
Afte rfinally getting gitversion working, I am ending up with version tags that look like "SemVer": "0.0.0-0-6-preview.1" - which by all accounts is not a valid semver.

## Expected Behavior
0.6-preview.1 as generated semver

## Actual Behavior
0.0.0-0-6-preview.1

## Possible Fix
No idea

## Steps to Reproduce
The following is my gitversion:

assembly-versioning-scheme: MajorMinorPatch
assembly-file-versioning-scheme: MajorMinorPatch
mode: ContinuousDelivery
label-prefix: '[vV]'
major-version-bump-message: '\+semver:\s?(breaking|major)'
minor-version-bump-message: '\+semver:\s?(feature|minor)'
patch-version-bump-message: '\+semver:\s?(fix|patch)'
no-bump-message: '\+semver:\s?(none|skip)'
label-pre-release-weight: 60000
commit-message-incrementing: Enabled
branches:
  develop:
    mode: ContinuousDeployment
    label: 'ci'
    increment: None
    prevent-increment-of-merged-branch-version: false
    track-merge-target: true
    regex: ^dev[/-]
    source-branches: []
    tracks-release-branches: true
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 0
  main:
    label: 'dev'
    increment: Patch
    prevent-increment-of-merged-branch-version: true
    track-merge-target: true
    regex: ^main$
    source-branches:
    - develop
    - feature
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: true
    pre-release-weight: 55000
  release:
    label: useBranchName
    increment: None
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
    regex: ^rel[/-]
    source-branches:
    - develop
    - main
    - support
    - release
    tracks-release-branches: false
    is-release-branch: true
    is-mainline: false
    pre-release-weight: 30000
  feature:
    label: useBranchName
    increment: Inherit
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    regex: ^ftr[/-]
    source-branches:
    - develop
    - main
    - release
    - feature
    - support
    - hotfix
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  pull-request:
    label: 'pr'
    increment: Inherit
    prevent-increment-of-merged-branch-version: false
    label-number-pattern: '[/-](?<number>\d+)'
    track-merge-target: false
    regex: ^req[/-]
    source-branches:
    - develop
    - main
    - release
    - feature
    - support
    - hotfix
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  hotfix:
    label: 'fix'
    increment: Patch
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    regex: ^fix[/-]
    source-branches:
    - develop
    - main
    - support
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  support:
    label: 'bug'
    increment: Patch
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
    regex: ^bug[/-]
    source-branches:
    - main
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: true
    pre-release-weight: 55000
ignore:
  sha: []
commit-date-format: yyyy-MM-dd
merge-message-formats: {}
update-build-number: true

The Azure Devops looks like:

Starting: GitVersion Execute
==============================================================================
Task         : Execute GitVersion Task
Description  : Easy Semantic Versioning (http://semver.org/) for projects using Git
Version      : 0.10.2
Author       : GitTools Contributors
Help         : See the [documentation](https://gitversion.net/docs/) for help
==============================================================================
Agent: 'Azure Pipelines'
Command: dotnet-gitversion D:/a/1/Repository /output json /output buildserver /config D:\a\1\s\pipelines\common\gitversion.yml /verbosity Diagnostic
C:\hostedtoolcache\windows\GitVersion.Tool\6.0.0-beta.2\x64\dotnet-gitversion.exe D:/a/1/Repository /output json /output buildserver /config D:\a\1\s\pipelines\common\gitversion.yml /verbosity Diagnostic
INFO [06/23/23 18:23:08:10] Working directory: D:/a/1/Repository
INFO [06/23/23 18:23:08:11] Project root is: D:\a\1\Repository\
INFO [06/23/23 18:23:08:11] DotGit directory is: D:\a\1\Repository\.git
INFO [06/23/23 18:23:08:11] Branch from build environment: refs/heads/rel/0.6-preview
INFO [06/23/23 18:23:08:11] Begin: Normalizing git directory for branch 'refs/heads/rel/0.6-preview'
  INFO [06/23/23 18:23:08:14] One remote found (origin -> 'https://nettecture@dev.azure.com/nettecture/Sandbox/_git/GitVersion').
  INFO [06/23/23 18:23:08:14] Skipping fetching, if GitVersion does not calculate your version as expected you might need to allow fetching or use dynamic repositories
  INFO [06/23/23 18:23:08:15] Creating local branch rel/0.6-preview
  INFO [06/23/23 18:23:08:19] Creating local branch from remote tracking 'refs/remotes/origin/develop'.
  INFO [06/23/23 18:23:08:20] Creating local branch from remote tracking 'refs/remotes/origin/main'.
  INFO [06/23/23 18:23:08:21] Creating local branch from remote tracking 'refs/remotes/origin/release/0.2'.
  INFO [06/23/23 18:23:08:21] Creating local branch from remote tracking 'refs/remotes/origin/release/0.3-preview'.
  INFO [06/23/23 18:23:08:21] Creating local branch from remote tracking 'refs/remotes/origin/release/0.4-preview'.
  INFO [06/23/23 18:23:08:21] Creating local branch from remote tracking 'refs/remotes/origin/release/0.5-preview'.
  INFO [06/23/23 18:23:08:22] HEAD points at branch 'refs/heads/rel/0.6-preview'.
  INFO [06/23/23 18:23:08:22] End: Normalizing git directory for branch 'refs/heads/rel/0.6-preview' (Took: 113.41ms)
  INFO [06/23/23 18:23:08:23] Begin: Loading version variables from disk cache
    INFO [06/23/23 18:23:08:23] Cache file D:\a\1\Repository\.git\gitversion_cache\9AC7CFCF791E396601AA1631E2668E04AB1D5279.yml not found.
  INFO [06/23/23 18:23:08:23] End: Loading version variables from disk cache (Took: 0.56ms)
  INFO [06/23/23 18:23:08:23] Using latest commit on specified branch
  INFO [06/23/23 18:23:08:45] Running against branch: rel/0.6-preview (54e392a WIP)
  INFO [06/23/23 18:23:08:45] Begin: Calculating the base versions
    INFO [06/23/23 18:23:08:47] Begin: Getting tagged semantic versions. LabelPrefix: [vV] and Format: Strict
    INFO [06/23/23 18:23:08:47] End: Getting tagged semantic versions. LabelPrefix: [vV] and Format: Strict (Took: 1.60ms)
    INFO [06/23/23 18:23:08:48] Using current branch name to calculate version tag
    INFO [06/23/23 18:23:08:50] Base version used: Fallback base version: 0.0.0 with commit count source External Source
  INFO [06/23/23 18:23:08:50] End: Calculating the base versions (Took: 55.69ms)
  INFO [06/23/23 18:23:08:51] Begin: Using manual deployment workflow to calculate the incremented version.
    INFO [06/23/23 18:23:08:51] Begin: Getting tagged semantic versions from 'refs/heads/rel/0.6-preview'.  LabelPrefix: [vV] and Format: Strict
      DEBUG [06/23/23 18:23:08:51] Returning cached tagged semantic versions. LabelPrefix: [vV] and Format: Strict
    INFO [06/23/23 18:23:08:51] End: Getting tagged semantic versions from 'refs/heads/rel/0.6-preview'.  LabelPrefix: [vV] and Format: Strict (Took: 1.46ms)
    INFO [06/23/23 18:23:08:51] 38 commits found between  and 54e392a WIP
  INFO [06/23/23 18:23:08:51] End: Using manual deployment workflow to calculate the incremented version. (Took: 5.07ms)
  INFO [06/23/23 18:23:08:51] Begin: Creating dictionary
  INFO [06/23/23 18:23:08:52] End: Creating dictionary (Took: 0.93ms)
  INFO [06/23/23 18:23:08:52] Begin: Storing version variables to cache file D:\a\1\Repository\.git\gitversion_cache\9AC7CFCF791E396601AA1631E2668E04AB1D5279.yml
  INFO [06/23/23 18:23:08:52] End: Storing version variables to cache file D:\a\1\Repository\.git\gitversion_cache\9AC7CFCF791E396601AA1631E2668E04AB1D5279.yml (Took: 0.99ms)
Executing GenerateSetVersionMessage for 'AzurePipelines'.
Executing GenerateBuildLogOutput for 'AzurePipelines'.
{
  "AssemblySemFileVer": "0.0.0.0",
  "AssemblySemVer": "0.0.0.0",
  "BranchName": "rel/0.6-preview",
  "BuildMetaData": 38,
  "CommitDate": "2023-06-23",
  "CommitsSinceVersionSource": 38,
  "EscapedBranchName": "rel-0-6-preview",
  "FullBuildMetaData": "38.Branch.rel-0.6-preview.Sha.54e392a30581905fd3d026a6aac2e2c69fd35d21",
  "FullSemVer": "0.0.0-0-6-preview.1+38",
  "InformationalVersion": "0.0.0-0-6-preview.1+38.Branch.rel-0.6-preview.Sha.54e392a30581905fd3d026a6aac2e2c69fd35d21",
  "Major": 0,
  "MajorMinorPatch": "0.0.0",
  "Minor": 0,
  "Patch": 0,
  "PreReleaseLabel": "0-6-preview",
  "PreReleaseLabelWithDash": "-0-6-preview",
  "PreReleaseNumber": 1,
  "PreReleaseTag": "0-6-preview.1",
  "PreReleaseTagWithDash": "-0-6-preview.1",
  "SemVer": "0.0.0-0-6-preview.1",
  "Sha": "54e392a30581905fd3d026a6aac2e2c69fd35d21",
  "ShortSha": "54e392a",
  "UncommittedChanges": 0,
  "VersionSourceSha": "",
  "WeightedPreReleaseNumber": 30001
}
  INFO [06/23/23 18:23:08:96] Done writing 
Async Command Start: Update Build Number
Update build number to 0.0.0-0-6-preview.1+38 for build 757
Async Command End: Update Build Number
Finishing: GitVersion Execute
## Your Environment
Azure Devops. This is a dummy project for setting up gitversion.

Installed version:

Starting: GitVersion Setup
==============================================================================
Task         : Setup GitVersion Task
Description  : Easy Semantic Versioning (http://semver.org/) for projects using Git
Version      : 0.10.2
Author       : GitTools Contributors
Help         : See the [documentation](https://gitversion.net/docs/) for help
==============================================================================
Agent: 'Azure Pipelines'

--------------------------
Acquiring GitVersion.Tool version spec: 6.x
--------------------------
Command: dotnet tool install GitVersion.Tool --tool-path D:\a\_temp --version 6.0.0-beta.2
"C:\Program Files\dotnet\dotnet.exe" tool install GitVersion.Tool --tool-path D:\a\_temp --version 6.0.0-beta.2
You can invoke the tool using the following command: dotnet-gitversion
Tool 'gitversion.tool' (version '6.0.0-beta.2') was successfully installed.
Caching tool: GitVersion.Tool 6.0.0-beta.2 x64
--------------------------
GitVersion.Tool version: 6.0.0-beta.2 installed.
--------------------------
Prepending PATH environment variable with directory: C:\hostedtoolcache\windows\GitVersion.Tool\6.0.0-beta.2\x64
Finishing: GitVersion Setup

**Comment by HHobeck** [comment]
Hi there. I'm going to move this issue to the Question/Answer section because the description seems to me not a bug report.

**Comment by HHobeck** [comment, marked as answer]
From my understanding git version does exactly what you have specified. You are ending up with this semantic version because:

1. The semantic-version-format has been set to Strict
1. The label-prefix has been set to [vV]
1. The label has been set to {BranchName} aka useBranchName
1. The increment has been set to None

Customizing workflows is a task which should be only an option if you have advanced experience with git version. For the beginning I would recommend you to use the standard supported git hub or git flow workflow: https://gitversion.net/docs/learn/branching-strategies/gitflow/examples

Happy branching!

Edit:
{major}. -> 0.
{minor}. -> 0.
{patch}- -> 0-
{pre-release-label}. -> 0-6-preview.
{pre-release-number} -> 1

**BTW the value 0.0.0-0-6-preview.1 is a valid semantic version**
Link [1]: [Semantic Versioning 2.0.0](https://semver.org/lang/de/)
Link [2]: [regex for semantic versioning](https://ihateregex.io/expr/semver/)

**Comment by NetTecture** [comment]
Still a bugt . I do not say yo uayre wrong - I am saying that if there is a strict semver and you do not find it, there should be an entry in the log that is easy to see.

I am so stuck it is not fun, and I can imagine I am the only one.

Ok, I got a little further, but the way this system insists on bothing up the version numbers is not fun.

* As per your statement and my config, I am ok with a branch that is i.e. /rel/v0.1.1 - great.
* COming directly from main in an empty repository for this, I do that and I get...
INFO [06/26/23 15:57:05:18] Begin: Normalizing git directory for branch 'refs/heads/rel/v0.1.0'
  INFO [06/26/23 15:57:05:21] One remote found (origin -> 'https://nettecture@dev.azure.com/nettecture/CoreLibraries/_git/CodeCore').
  INFO [06/26/23 15:57:05:21] Skipping fetching, if GitVersion does not calculate your version as expected you might need to allow fetching or use dynamic repositories
  INFO [06/26/23 15:57:05:22] Creating local branch rel/v0.1.0
  INFO [06/26/23 15:57:05:25] Creating local branch from remote tracking 'refs/remotes/origin/main'.
  INFO [06/26/23 15:57:05:26] HEAD points at branch 'refs/heads/rel/v0.1.0'.
  INFO [06/26/23 15:57:05:26] End: Normalizing git directory for branch 'refs/heads/rel/v0.1.0' (Took: 85.85ms)
  INFO [06/26/23 15:57:05:27] Begin: Loading version variables from disk cache
    INFO [06/26/23 15:57:05:27] Cache file D:\a\1\Repository\.git\gitversion_cache\C47BB79796A4ABC93F6F82DF99413B6C1334E303.yml not found.
  INFO [06/26/23 15:57:05:27] End: Loading version variables from disk cache (Took: 0.56ms)
  INFO [06/26/23 15:57:05:27] Using latest commit on specified branch
  INFO [06/26/23 15:57:05:48] Running against branch: rel/v0.1.0 (a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore)
  INFO [06/26/23 15:57:05:49] Begin: Calculating the base versions
    INFO [06/26/23 15:57:05:52] Begin: Getting tagged semantic versions. TagPrefix: [vV] and Format: Strict
    INFO [06/26/23 15:57:05:52] End: Getting tagged semantic versions. TagPrefix: [vV] and Format: Strict (Took: 1.20ms)
    INFO [06/26/23 15:57:05:55] Begin: Finding branch source of 'rel/v0.1.0'
      INFO [06/26/23 15:57:05:55] Begin: Finding merge base between 'rel/v0.1.0' and 'main'.
        INFO [06/26/23 15:57:05:56] Found merge base of a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
        INFO [06/26/23 15:57:05:56] Merge base of rel/v0.1.0' and 'main is a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
      INFO [06/26/23 15:57:05:56] End: Finding merge base between 'rel/v0.1.0' and 'main'. (Took: 8.52ms)
      INFO [06/26/23 15:57:05:56] Begin: Finding merge base between 'rel/v0.1.0' and 'origin/main'.
        INFO [06/26/23 15:57:05:56] Found merge base of a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
        INFO [06/26/23 15:57:05:56] Merge base of rel/v0.1.0' and 'origin/main is a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
      INFO [06/26/23 15:57:05:56] End: Finding merge base between 'rel/v0.1.0' and 'origin/main'. (Took: 0.72ms)
      INFO [06/26/23 15:57:05:56] Begin: Finding merge base between 'rel/v0.1.0' and 'origin/rel/v0.1.0'.
        INFO [06/26/23 15:57:05:56] Found merge base of a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
        INFO [06/26/23 15:57:05:57] Merge base of rel/v0.1.0' and 'origin/rel/v0.1.0 is a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
      INFO [06/26/23 15:57:05:57] End: Finding merge base between 'rel/v0.1.0' and 'origin/rel/v0.1.0'. (Took: 0.50ms)
      INFO [06/26/23 15:57:05:57] Multiple source branches have been found, picking the first one (main).
This may result in incorrect commit counting.
Options were:
main, origin/main
    INFO [06/26/23 15:57:05:57] End: Finding branch source of 'rel/v0.1.0' (Took: 21.55ms)
    INFO [06/26/23 15:57:05:57] Version in branch name: 0.1.0 with commit count source a68e0fc5a46d01de384ffa7f822d4f3310010f8b
    INFO [06/26/23 15:57:05:57] Using current branch name to calculate version tag
    INFO [06/26/23 15:57:05:58] Found multiple base versions which will produce the same SemVer (0.1.0-v0-1-0.1), taking oldest source for commit counting (Version in branch name)
    INFO [06/26/23 15:57:05:58] Base version used: Version in branch name: 0.1.0 with commit count source a68e0fc5a46d01de384ffa7f822d4f3310010f8b
  INFO [06/26/23 15:57:05:58] End: Calculating the base versions (Took: 99.01ms)
  INFO [06/26/23 15:57:05:59] Begin: Using manual deployment workflow to calculate the incremented version.
    INFO [06/26/23 15:57:05:59] Begin: Getting tagged semantic versions from 'refs/heads/rel/v0.1.0'.  TagPrefix: [vV] and Format: Strict
      DEBUG [06/26/23 15:57:05:59] Returning cached tagged semantic versions. TagPrefix: [vV] and Format: Strict
    INFO [06/26/23 15:57:05:59] End: Getting tagged semantic versions from 'refs/heads/rel/v0.1.0'.  TagPrefix: [vV] and Format: Strict (Took: 0.93ms)
    INFO [06/26/23 15:57:05:59] 0 commits found between a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore and a68e0fc Merge branch 'main' of https://dev.azure.com/nettecture/CoreLibraries/_git/CodeCore
  INFO [06/26/23 15:57:05:59] End: Using manual deployment workflow to calculate the incremented version. (Took: 3.51ms)
  INFO [06/26/23 15:57:05:59] Begin: Creating dictionary
  INFO [06/26/23 15:57:05:59] End: Creating dictionary (Took: 0.91ms)
  INFO [06/26/23 15:57:05:59] Begin: Storing version variables to cache file D:\a\1\Repository\.git\gitversion_cache\C47BB79796A4ABC93F6F82DF99413B6C1334E303.yml
  INFO [06/26/23 15:57:05:59] End: Storing version variables to cache file D:\a\1\Repository\.git\gitversion_cache\C47BB79796A4ABC93F6F82DF99413B6C1334E303.yml (Took: 0.86ms)

This looks good. It finds 0.1.0 - AMAZING. Except then I read....

Executing GenerateBuildLogOutput for 'AzurePipelines'.
{
  "AssemblySemFileVer": "0.1.0.0",
  "AssemblySemVer": "0.1.0.0",
  "BranchName": "rel/v0.1.0",
  "BuildMetaData": 0,
  "CommitDate": "2023-06-26",
  "CommitsSinceVersionSource": 0,
  "EscapedBranchName": "rel-v0-1-0",
  "FullBuildMetaData": "0.Branch.rel-v0.1.0.Sha.a68e0fc5a46d01de384ffa7f822d4f3310010f8b",
  "FullSemVer": "0.1.0-v0-1-0.1+0",
  "InformationalVersion": "0.1.0-v0-1-0.1+0.Branch.rel-v0.1.0.Sha.a68e0fc5a46d01de384ffa7f822d4f3310010f8b",
  "Major": 0,
  "MajorMinorPatch": "0.1.0",
  "Minor": 1,
  "Patch": 0,
  "PreReleaseLabel": "v0-1-0",
  "PreReleaseLabelWithDash": "-v0-1-0",
  "PreReleaseNumber": 1,
  "PreReleaseTag": "v0-1-0.1",
  "PreReleaseTagWithDash": "-v0-1-0.1",
  "SemVer": "0.1.0-v0-1-0.1",
  "Sha": "a68e0fc5a46d01de384ffa7f822d4f3310010f8b",
  "ShortSha": "a68e0fc",
  "UncommittedChanges": 0,
  "VersionSourceSha": "a68e0fc5a46d01de384ffa7f822d4f3310010f8b",
  "WeightedPreReleaseNumber": 30001
}

See the full sever? Somehow "AssemblySemFileVer": "0.1.0.0", leads to... "FullSemVer": "0.1.0-v0-1-0.1+0",

Ok, there is no easy documentation for 6.0 (yet - get that), but thi is close to frustrating. I would have expected that to result in a 0.1.1 versoin number WITHOUT the repeat.

**Reply by HHobeck** [reply]
Sorry if I'm bothering you but this is not a new feature. Again if you are configuring that the pre release label name needs to be the name of the branch then the branch name will be used as the pre release label name. In your case the branch name is rel/v0.1.0. I think a lot of people who are using the label configuration knowing this.

If you don't want to generate a pre-release version you need to set the label especially in the 6.0.0-beta.2 version to null (please ensure to do this on the branch and on the root level).

**In advance I want to tell you:**
that I am very surprised by your criticism and the way you are express it. Do you know that:

- a beta version is a version which is not finished?
- this is an open source project and you are able to take a look in to the source code?
- this project will be written and maintained from people in there spare time without payment?
- you are able to help us improving this project with creating well written and analysed issues?
- you are able making changes and creating pull requests?
